packages/semaphore/src/Semaphore.ts

```diff
@@ -0,0 +1,43 @@
+export class Permit {
+  private onRelease: () => void;
+  private isReleased = false;
+
+  constructor(onRelease: () => void) {
+    this.onRelease = onRelease;
+  }
+
+  release() {
+    if (!this.isReleased) {
+      this.isReleased = true;
+      this.onRelease();
+    }
+  }
+}
+
+export class Semaphore {
+  private availablePermits: number;
+  private deferreds: {resolve(permit: Permit): void}[] = [];
+
+  constructor(count: number) {
+    this.availablePermits = count;
+  }
+
+  acquire(): Promise<Permit> {
+    if (this.availablePermits > 0) {
+      this.availablePermits--;
+      return Promise.resolve(this.createPermit());
+    } else {
+      return new Promise(resolve => this.deferreds.push({resolve}));
+    }
+  }
+
+  private createPermit(): Permit {
+    return new Permit(() => {
+      this.availablePermits++;
+
+      if (this.deferreds.length) {
+        this.deferreds.shift()!.resolve(this.createPermit());
```

Comment:
Would you mind explaining this line? Also, I'm noticing the use of `!` which would basically ignore or bypass typechecking for `null` or `undefined`. It's because you're checking `.length` right before?

Reply:
"If there are pending requests to acquire a permit, pop the first one (first in is first out) and resolve the pending promise of that request."

Yes, `!` is to silence TypeScript since I'm already checking length.